brambling/views/organizer.py

```diff
@@ -1182,29 +1186,75 @@ def get_queryset(self):
         )
 
 
-class RefundView(View):
+class RefundView(FormView):
+    form_class = TransactionRefundForm
+    template_name = "brambling/event/organizer/refund_form.html"
+
+    def get_context_data(self, **kwargs):
+        context = super(RefundView, self).get_context_data(**kwargs)
+        context['event'] = self.event
+        context['order'] = self.transaction.order
+        context['organization'] = self.event.organization
+        context['transaction'] = self.transaction
+        return context
+
     def get_object(self):
         self.event = get_object_or_404(Event.objects.select_related('organization'),
                                        slug=self.kwargs['event_slug'],
                                        organization__slug=self.kwargs['organization_slug'])
         if not self.request.user.has_perm('edit', self.event):
             raise Http404
         try:
-            return Transaction.objects.get(
+            self.transaction = Transaction.objects.get(
                 event=self.event,
                 order__code=self.kwargs['code'],
                 pk=self.kwargs['pk']
             )
+            return self.transaction
         except Transaction.DoesNotExist:
             raise Http404
 
-    def post(self, request, *args, **kwargs):
-        txn = self.get_object()
+    def get_form_kwargs(self):
+        kwargs = super(RefundView, self).get_form_kwargs()
+        kwargs['transaction'] = self.get_object()
+        return kwargs
+
+    def form_valid(self, form):
+        transaction = self.transaction
+        # Must be an empty queryset if no items to refund.
+        # If transaction.refund receives `None` it will refund all items.
+        bought_items = form.cleaned_data.get('items', BoughtItem.objects.none())
+
+        # Compare to None to allow for an empty queryset
+        if bought_items is not None:
+            bought_items = BoughtItem.objects.filter(pk__in=[x.pk for x in bought_items])
```

Comment:
The `BoughtItem.objects.none()` bit can actually be deleted. This is still necessary because form data returns a list of instances instead of a queryset. (`bought_items.update()` is called in the `.refund()` method)